Add unit tests intmatrix

RichardWaiteSTFC opened the pull request.

rebeccafair reviewed (changes requested):
Looks good! Thanks for fixing the symbolic/biquadratic issues. Just a few more things we could try:

* Improve docstring e.g. currently it implies `SS` contains all fields including `aniso`, `g` etc.
* Test with no `g`
* Test with no `aniso`
* Test with `symbolic` and `fitmode` off
* Test with 2 atoms/cell
* Test with a different `sym`

Some of those could probably go in the same test. Also I think there's a merge conflict with the changelog.

RichardWaiteSTFC commented:
> * Test with no `g`
> * Test with no `aniso`

I believe these are covered by `test_intmatrix_no_couplings_defined` - or do you mean with some non-zero exchanges?

> * Test with 2 atoms/cell
> * Test with a different `sym`

There are 2 atoms in the unit cell in these tests as one is generated by the body centered spacegroup - however I could make a test where I revert to $P1$ symmetry and add the atom at (0.5, 0.5, 0.5) with `addatom` - would that work?

rebeccafair commented:
> > * Test with no `g`
> > * Test with no `aniso`
> 
> I believe these are covered by `test_intmatrix_no_couplings_defined` - or do you mean with some non-zero exchanges?

Ah sorry! I meant a case where a magnetic atom is defined, but no `aniso` or `g` are added, to trigger lines 149 and 162

> > * Test with 2 atoms/cell
> > * Test with a different `sym`
> 
> There are 2 atoms in the unit cell in these tests as one is generated by the body centered spacegroup - however I could make a test where I revert to P1 symmetry and add the atom at (0.5, 0.5, 0.5) with `addatom` - would that work?

That sounds sensible!

RichardWaiteSTFC commented:
Thanks for the review, should be ready for re-review whenever you get the chance!

> * Improve docstring e.g. currently it implies `SS` contains all fields including `aniso`, `g` etc.

Done - hopefully it's clearer. I think when we move over to python `SS` should actually be split into a couple of structs (or even it's own class) so the output dimensions, fields and contents are more consistent and not dependent on the input parameters.

> * Test with no `g` 
> * Test with no `aniso`

Added a test with no `g` and `aniso` with an matom defined to trigger those lines I had missed

> * Test with `symbolic` and `fitmode` off

Done

> * Test with 2 atoms/cell
> * Test with a different `sym`

Added test with P 1 symmetry and manually added the body-centered atom. There was also an existing test that had a face-centred spacegroup `test_sortDM_reorders_bonds`.

rebeccafair reviewed (approved):
Thanks for the changes, looks good to me!

mducle commented on swfiles/@spinw/intmatrix.m:
```diff
@@ -62,31 +62,38 @@
+%   Subsequent rows in these matrices are the elements of the 3 x 3 
+%   exchange matrix `[Jxx; Jxy; Jxz; Jyx; Jyy; Jyz; Jzx; Jzy; Jzz]`
+%   and the final row indicates whether the coupling is
+%   quadratic (0) or biquadratic (1). The `dip` field contains the dipolar
+%   interactions only that are not added to `SS.all.
```
I think `quadratic (0)` here should be `bilinear (0)`. The "bi" refers to pair-interactions between spins; `Ji.Jj` is a linear pair interaction whilst `(Ji.Jj)^2` is a quadratic interaction.

Also, you're missing closing &#96; after `SS.all`.

mducle commented on swfiles/@spinw/intmatrix.m:
```diff
@@ -62,31 +62,38 @@
+%  * `dm`  : Subsequent rows contain DM interaction `[DMx; DMy; DMz]`
+%  * `gen` : Subsequent rows contain a general interaction 
+%            `[Jxx; Jxy; Jxz; Jyx; Jyy; Jyz; Jzx; Jzy; Jzz]`
+%  * `bq   : One subsequent row which is the isotropic exchane as in SS.iso
```
Missing &#96; after `bq` and a typo: `exchane`

mducle reviewed (approved):
Looks, good... just a couple of minor typos in the docstrings.
